pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go

```diff
@@ -0,0 +1,241 @@
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "deletion failed, would retry exponentially", "nsxserviceaccount", req.NamespacedName)
+				deleteFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("removed finalizer", "nsxserviceaccount", req.NamespacedName)
+			deleteSuccess(r, &ctx, obj)
+		} else {
+			// only print a message because it's not a normal case
+			log.Info("finalizers cannot be recognized", "nsxserviceaccount", req.NamespacedName)
+		}
+	}
+
+	return ResultNormal, nil
+}
+
+// setupWithManager sets up the controller with the Manager.
+func (r *NSXServiceAccountReconciler) setupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&nsxvmwarecomv1alpha1.NSXServiceAccount{}).
+		WithEventFilter(predicate.Funcs{
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				// Suppress Delete events to avoid filtering them out in the Reconcile function
+				return false
+			},
+		}).
+		WithOptions(
+			controller.Options{
+				MaxConcurrentReconciles: runtime.NumCPU(),
+			}).
+		Complete(r)
+}
+
+// Start setup manager and launch GC
+func (r *NSXServiceAccountReconciler) Start(mgr ctrl.Manager) error {
+	err := r.setupWithManager(mgr)
+	if err != nil {
+		return err
+	}
+
+	go r.GarbageCollector(make(chan bool), servicecommon.GCInterval)
+	return nil
+}
+
+// GarbageCollector collect NSXServiceAccount which has been removed from crd.
+// cancel is used to break the loop during UT
+func (r *NSXServiceAccountReconciler) GarbageCollector(cancel chan bool, timeout time.Duration) {
+	ctx := context.Background()
+	log.Info("garbage collector started")
+	for {
+		select {
+		case <-cancel:
+			return
+		case <-time.After(timeout):
+		}
+		nsxServiceAccountUIDSet := r.Service.ListNSXServiceAccountRealization()
+		if len(nsxServiceAccountUIDSet) == 0 {
+			continue
+		}
+		nsxServiceAccountList := &nsxvmwarecomv1alpha1.NSXServiceAccountList{}
+		err := r.Client.List(ctx, nsxServiceAccountList)
+		if err != nil {
+			log.Error(err, "failed to list NSXServiceAccount CR")
+			continue
+		}
+		r.garbageCollector(nsxServiceAccountUIDSet, nsxServiceAccountList)
+	}
+}
+
+func (r *NSXServiceAccountReconciler) garbageCollector(nsxServiceAccountUIDSet sets.String, nsxServiceAccountList *nsxvmwarecomv1alpha1.NSXServiceAccountList) (gcSuccessCount, gcErrorCount uint32) {
+	nsxServiceAccountCRUIDMap := map[string]types.NamespacedName{}
+	for _, nsxServiceAccount := range nsxServiceAccountList.Items {
+		nsxServiceAccountCRUIDMap[string(nsxServiceAccount.UID)] = types.NamespacedName{
+			Namespace: nsxServiceAccount.Namespace,
+			Name:      nsxServiceAccount.Name,
+		}
+	}
+
+	for nsxServiceAccountUID := range nsxServiceAccountUIDSet {
+		if _, ok := nsxServiceAccountCRUIDMap[nsxServiceAccountUID]; ok {
+			continue
+		}
+		log.V(1).Info("GC collected NSXServiceAccount CR", "UID", nsxServiceAccountUID)
```

Comment:
Maybe "collect" is better than "collected", because at this time it hasn't really deletes the resources yet.

Reply:
collected -> collects